src/Microsoft.Diagnostics.EventFlow.Core/Implementations/LossReportingPropagatorBlock.cs

```diff
@@ -0,0 +1,209 @@
+﻿// ------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All rights reserved.
+//  Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Validation;
+
+namespace Microsoft.Diagnostics.EventFlow
+{
+    /// <summary>
+    /// The purpose of this class is to ensure that EventFlow pipeline does not drop events without warning through the health reporter.
+    /// This is a possiblity with BroadcastBlock, which, by design, will discard the current message once a new message arrives and the 
+    /// current message has been offered to all downstream blocks. This happens regardless whether the message was accepted, postponed
+    /// or discarded by the downstream block. 
+    /// The LossReportingPropagatorBlock works around this issue by reporting a message loss when a passing message is discarded by the downstream block,
+    /// or when a message is postponed and then not available for consumption later on.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal class LossReportingPropagatorBlock<T> : IPropagatorBlock<T, T>
+    {
+        private IHealthReporter healthReporter;
+        private ITargetBlock<T> target;
+        private ISourceBlock<T> source;
+        private bool propagateCompletion;
+        private bool completed;
+        private Lazy<TaskCompletionSource<bool>> completionSource;
+
+        public LossReportingPropagatorBlock(IHealthReporter healthReporter)
+        {
+            Requires.NotNull(healthReporter, nameof(healthReporter));
+            this.healthReporter = healthReporter;
+            this.target = null;
+            this.propagateCompletion = false;
+            this.completed = false;
+            this.completionSource = new Lazy<TaskCompletionSource<bool>>(LazyThreadSafetyMode.PublicationOnly);
+        }
+
+        public Task Completion
+        {
+            get
+            {
+                if (this.propagateCompletion)
```

Comment:
Any reason passing the CompletionTask from the target block when propagateCompletion? Remember the bufferBlock will complete regardless of the subsequent's completion task. I think it's true for other blocks. Although it doesn't matter right now, it's better to be consistent. And we can get rid of the this.completed and only use one tcs.